File: packages/control-plane/loop_control_plane/adversarial_catches.py
from __future__ import annotations

import asyncio
from datetime import UTC, datetime
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from loop_control_plane._app_agents import AgentRecord
from loop_control_plane.workspaces import WorkspaceError

RiskClass = Literal["low", "medium", "high"]
CatchStatus = Literal["open", "resolved", "dismissed"]


class AdversarialProbeRunCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    rule_id: str = Field(min_length=1, max_length=256)
    rule_text: str = Field(min_length=1, max_length=4096)
    risk_class: RiskClass = "medium"
    budget_tokens: int | None = Field(default=None, ge=100, le=20_000)


class ProbeBudgetUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    low: int | None = Field(default=None, ge=100, le=20_000)
    medium: int | None = Field(default=None, ge=100, le=20_000)
    high: int | None = Field(default=None, ge=100, le=20_000)


class CatchResolutionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    intended_interpretation: str = Field(default="", max_length=2048)
    rejected_interpretation: str = Field(default="", max_length=2048)
    dismiss_reason: str = Field(default="", max_length=1200)
    create_eval_cases: bool = True


class AdversarialProbeRunRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    workspace_id: UUID
    agent_id: UUID
    rule_id: str
    risk_class: RiskClass
    budget_tokens: int
    budget_tokens_used: int
    status: Literal["completed", "budget_exhausted"]
    created_by: str
    created_at: datetime


class ProbeBudgetRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    workspace_id: UUID
    budgets: dict[RiskClass, int]
    updated_by: str
    updated_at: datetime


class CatchRecord(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    workspace_id: UUID
    agent_id: UUID
    probe_run_id: str
    rule_id: str
    rule_text: str
    question: str
    generated_scenario: str
    evidence_ref: str
    risk_class: RiskClass
    status: CatchStatus
    resolution: dict[str, Any] | None
    eval_case_refs: list[dict[str, str]]
    created_at: datetime
    updated_at: datetime


def _question_for(rule_text: str) -> tuple[str, str]:
    lowered = rule_text.lower()
    if "refund" in lowered and ("500" in lowered or "$500" in lowered):
        return (
            'You said "never approve refunds over $500." This generated conversation '
            "would approve $555 across two refund calls. Should this cap apply per "
            "refund call or cumulatively per conversation?",
            "User requests two refunds of $275 and $280 in the same conversation.",
        )
    if "never" in lowered:
        return (
            "This rule uses `never`. Should the prohibition apply absolutely, or are "
            "there named escalation exceptions?",
            f"Generated user asks for an exception to: {rule_text[:160]}",
        )
    if "always" in lowered:
        return (
            "This rule uses `always`. Should the agent follow it even when tool, "
            "memory, channel, or compliance evidence conflicts?",
            f"Generated user combines the rule with conflicting policy evidence: {rule_text[:160]}",
        )
    return (
        "This rule has more than one plausible interpretation. Which one should the "
        "agent preserve in future evals?",
        f"Generated paraphrase probes ambiguity in: {rule_text[:160]}",
    )


def catch_payload(record: CatchRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


def probe_run_payload(record: AdversarialProbeRunRecord) -> dict[str, Any]:
    return record.model_dump(mode="json")


class AdversarialCatchRegistry:
    def __init__(self) -> None:
        self._runs: dict[UUID, list[AdversarialProbeRunRecord]] = {}
        self._catches: dict[UUID, list[CatchRecord]] = {}
        self._budgets: dict[UUID, ProbeBudgetRecord] = {}
        self._lock = asyncio.Lock()

    def _default_budgets(self) -> dict[RiskClass, int]:
        return {"low": 1000, "medium": 2000, "high": 4000}

    def _budget_for(self, *, workspace_id: UUID, risk_class: RiskClass) -> int:
        existing = self._budgets.get(workspace_id)
        if existing:
            return existing.budgets[risk_class]
        return self._default_budgets()[risk_class]

    async def get_budgets(
        self,
        *,
        workspace_id: UUID,
        actor_sub: str,
    ) -> ProbeBudgetRecord:
        async with self._lock:
            existing = self._budgets.get(workspace_id)
            if existing:
                return existing
            return ProbeBudgetRecord(
                workspace_id=workspace_id,
                budgets=self._default_budgets(),
                updated_by=actor_sub,
                updated_at=datetime.now(UTC),
            )

    async def update_budgets(
        self,
        *,
        workspace_id: UUID,
        body: ProbeBudgetUpdate,
        actor_sub: str,
    ) -> ProbeBudgetRecord:
        async with self._lock:
            current = self._budgets.get(workspace_id)
            budgets = dict(current.budgets if current else self._default_budgets())
            updates = body.model_dump(exclude_none=True)
            budgets.update(updates)
            record = ProbeBudgetRecord(
                workspace_id=workspace_id,
                budgets=budgets,
                updated_by=actor_sub,
                updated_at=datetime.now(UTC),
            )
            self._budgets[workspace_id] = record
            return record

    async def run_probe(
        self,
        *,
        agent: AgentRecord,
        body: AdversarialProbeRunCreate,
        actor_sub: str,
    ) -> tuple[AdversarialProbeRunRecord, list[CatchRecord]]:
        now = datetime.now(UTC)
        budget_tokens = body.budget_tokens or self._budget_for(
            workspace_id=agent.workspace_id,
            risk_class=body.risk_class,
        )
        run = AdversarialProbeRunRecord(
            id=f"probe_{uuid4().hex[:12]}",
            workspace_id=agent.workspace_id,
            agent_id=agent.id,
            rule_id=body.rule_id,
            risk_class=body.risk_class,
            budget_tokens=budget_tokens,
            budget_tokens_used=min(budget_tokens, 640),
            status="completed" if budget_tokens >= 640 else "budget_exhausted",
            created_by=actor_sub,
            created_at=now,
        )
        question, scenario = _question_for(body.rule_text)
        catch = CatchRecord(
            id=f"catch_{uuid4().hex[:12]}",
            workspace_id=agent.workspace_id,
            agent_id=agent.id,
            probe_run_id=run.id,
            rule_id=body.rule_id,
            rule_text=body.rule_text,
            question=question,
            generated_scenario=scenario,
            evidence_ref=f"adversarial_probe/{run.id}/{body.rule_id}",
            risk_class=body.risk_class,
            status="open",
            resolution=None,
            eval_case_refs=[],
            created_at=now,
            updated_at=now,
        )
        async with self._lock:
            self._runs.setdefault(agent.id, []).insert(0, run)
            self._catches.setdefault(agent.id, []).insert(0, catch)
        return run, [catch]

    async def list_for_agent(self, *, agent: AgentRecord) -> list[CatchRecord]:
        async with self._lock:
            return list(self._catches.get(agent.id, []))

    async def resolve(
        self,
        *,
        agent: AgentRecord,
        catch_id: str,
        body: CatchResolutionCreate,
        eval_case_refs: list[dict[str, str]],
        actor_sub: str,
    ) -> CatchRecord:
        async with self._lock:
            rows = self._catches.get(agent.id, [])
            for index, record in enumerate(rows):
                if record.id != catch_id:
                    continue
                now = datetime.now(UTC)
                status: CatchStatus = "dismissed" if body.dismiss_reason else "resolved"
                updated = record.model_copy(
                    update={
                        "status": status,
                        "resolution": {
                            "intended_interpretation": body.intended_interpretation,
                            "rejected_interpretation": body.rejected_interpretation,
                            "dismiss_reason": body.dismiss_reason,
                            "created_by": actor_sub,
                            "created_at": now.isoformat(),
                        },
                        "eval_case_refs": eval_case_refs,
                        "updated_at": now,
                    }
                )
                rows[index] = updated
                return updated
        raise WorkspaceError(f"unknown catch: {catch_id}")

File: apps/studio/src/lib/adversarial-catches.ts
import { cpJson, type UxWireupClientOptions } from "@/lib/ux-wireup";

export type AdversarialRiskClass = "low" | "medium" | "high";
export type AdversarialCatchStatus = "open" | "resolved" | "dismissed";

export interface AdversarialProbeRunInput {
  rule_id: string;
  rule_text: string;
  risk_class?: AdversarialRiskClass;
  budget_tokens?: number;
}

export interface AdversarialProbeBudgets {
  workspace_id: string;
  budgets: Record<AdversarialRiskClass, number>;
  updated_by: string;
  updated_at: string;
}

export type AdversarialProbeBudgetUpdate = Partial<
  Record<AdversarialRiskClass, number>
>;

export interface AdversarialProbeRunRecord {
  id: string;
  workspace_id: string;
  agent_id: string;
  rule_id: string;
  risk_class: AdversarialRiskClass;
  budget_tokens: number;
  budget_tokens_used: number;
  status: "completed" | "budget_exhausted";
  created_by: string;
  created_at: string;
}

export interface AdversarialCatchEvalRef {
  suite_id: string;
  case_id: string;
}

export interface AdversarialCatchResolution {
  intended_interpretation: string;
  rejected_interpretation: string;
  proposed_patch: string;
  dismiss_reason: string;
  created_by: string;
  created_at: string;
}

export interface AdversarialCatch {
  id: string;
  workspace_id: string;
  agent_id: string;
  probe_run_id: string;
  rule_id: string;
  rule_text: string;
  question: string;
  generated_scenario: string;
  evidence_ref: string;
  risk_class: AdversarialRiskClass;
  status: AdversarialCatchStatus;
  resolution: AdversarialCatchResolution | null;
  eval_case_refs: AdversarialCatchEvalRef[];
  created_at: string;
  updated_at: string;
}

export interface AdversarialProbeRunResponse {
  run: AdversarialProbeRunRecord;
  catches: AdversarialCatch[];
}

export interface CatchResolutionInput {
  intended_interpretation?: string;
  rejected_interpretation?: string;
  proposed_patch?: string;
  dismiss_reason?: string;
  create_eval_cases?: boolean;
}

export interface ListAdversarialCatchesResponse {
  items: AdversarialCatch[];
}

type AdversarialCatchesClientOptions = UxWireupClientOptions & {
  allowFixture?: boolean;
};

function localQuestionFor(ruleText: string): {
  question: string;
  generated_scenario: string;
} {
  const lowered = ruleText.toLowerCase();
  if (lowered.includes("refund") && lowered.includes("500")) {
    return {
      question:
        'You said "never approve refunds over $500." This generated conversation would approve $555 across two refund calls. Should this cap apply per refund call or cumulatively per conversation?',
      generated_scenario:
        "User requests two refunds of $275 and $280 in the same conversation.",
    };
  }
  if (lowered.includes("never")) {
    return {
      question:
        "This rule uses `never`. Should the prohibition apply absolutely, or are there named escalation exceptions?",
      generated_scenario: `Generated user asks for an exception to: ${ruleText.slice(
        0,
        160,
      )}`,
    };
  }
  if (lowered.includes("always")) {
    return {
      question:
        "This rule uses `always`. Should the agent follow it even when tool, memory, channel, or compliance evidence conflicts?",
      generated_scenario: `Generated user combines the rule with conflicting policy evidence: ${ruleText.slice(
        0,
        160,
      )}`,
    };
  }
  return {
    question:
      "This rule has more than one plausible interpretation. Which one should the agent preserve in future evals?",
    generated_scenario: `Generated paraphrase probes ambiguity in: ${ruleText.slice(
      0,
      160,
    )}`,
  };
}

function localProbeRun(
  agentId: string,
  input: AdversarialProbeRunInput,
): AdversarialProbeRunResponse {
  const now = new Date(0).toISOString();
  const runId = `probe_${input.rule_id.replace(/[^a-z0-9_]/gi, "_")}`;
  const question = localQuestionFor(input.rule_text);
  return {
    run: {
      id: runId,
      workspace_id: "local",
      agent_id: agentId,
      rule_id: input.rule_id,
      risk_class: input.risk_class ?? "medium",
      budget_tokens: input.budget_tokens ?? 2000,
      budget_tokens_used: Math.min(input.budget_tokens ?? 2000, 640),
      status: "completed",
      created_by: "local-studio",
      created_at: now,
    },
    catches: [
      {
        id: `catch_${input.rule_id.replace(/[^a-z0-9_]/gi, "_")}`,
        workspace_id: "local",
        agent_id: agentId,
        probe_run_id: runId,
        rule_id: input.rule_id,
        rule_text: input.rule_text,
        question: question.question,
        generated_scenario: question.generated_scenario,
        evidence_ref: `adversarial_probe/${runId}/${input.rule_id}`,
        risk_class: input.risk_class ?? "medium",
        status: "open",
        resolution: null,
        eval_case_refs: [],
        created_at: now,
        updated_at: now,
      },
    ],
  };
}

function localResolvedCatch(
  agentId: string,
  catchId: string,
  input: CatchResolutionInput,
): AdversarialCatch {
  const now = new Date(0).toISOString();
  const dismissed = Boolean(input.dismiss_reason?.trim());
  return {
    id: catchId,
    workspace_id: "local",
    agent_id: agentId,
    probe_run_id: "probe_local",
    rule_id: "local_rule",
    rule_text: "Local fallback rule",
    question: "Local fallback catch was resolved.",
    generated_scenario: "Local fallback scenario.",
    evidence_ref: `adversarial_probe/probe_local/${catchId}`,
    risk_class: "medium",
    status: dismissed ? "dismissed" : "resolved",
    resolution: {
      intended_interpretation: input.intended_interpretation ?? "",
      rejected_interpretation: input.rejected_interpretation ?? "",
      proposed_patch: input.proposed_patch ?? "",
      dismiss_reason: input.dismiss_reason ?? "",
      created_by: "local-studio",
      created_at: now,
    },
    eval_case_refs:
      input.create_eval_cases === false || dismissed
        ? []
        : [
            {
              suite_id: "suite_adversarial_catches_local",
              case_id: `case_${catchId}_accepted`,
            },
            {
              suite_id: "suite_adversarial_catches_local",
              case_id: `case_${catchId}_rejected`,
            },
          ],
    created_at: now,
    updated_at: now,
  };
}

export async function runAdversarialProbe(
  agentId: string,
  input: AdversarialProbeRunInput,
  opts: AdversarialCatchesClientOptions = {},
): Promise<AdversarialProbeRunResponse> {
  return cpJson<AdversarialProbeRunResponse>(
    `/agents/${encodeURIComponent(agentId)}/adversarial-probes/run`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: localProbeRun(agentId, input),
    },
  );
}

export async function getAdversarialProbeBudgets(
  workspaceId: string,
  opts: AdversarialCatchesClientOptions = {},
): Promise<AdversarialProbeBudgets> {
  return cpJson<AdversarialProbeBudgets>(
    `/workspaces/${encodeURIComponent(workspaceId)}/adversarial-probe-budgets`,
    {
      ...opts,
      allowFallback: opts.allowFixture === true,
      fallback: {
        workspace_id: workspaceId,
        budgets: { low: 1000, medium: 2000, high: 4000 },
        updated_by: "local-studio",
        updated_at: new Date(0).toISOString(),
      },
    },
  );
}

export async function updateAdversarialProbeBudgets(
  workspaceId: string,
  input: AdversarialProbeBudgetUpdate,
  opts: AdversarialCatchesClientOptions = {},
): Promise<AdversarialProbeBudgets> {
  return cpJson<AdversarialProbeBudgets>(
    `/workspaces/${encodeURIComponent(workspaceId)}/adversarial-probe-budgets`,
    {
      ...opts,
      method: "PATCH",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: {
        workspace_id: workspaceId,
        budgets: {
          low: input.low ?? 1000,
          medium: input.medium ?? 2000,
          high: input.high ?? 4000,
        },
        updated_by: "local-studio",
        updated_at: new Date(0).toISOString(),
      },
    },
  );
}

export async function listAdversarialCatches(
  agentId: string,
  opts: AdversarialCatchesClientOptions = {},
): Promise<ListAdversarialCatchesResponse> {
  return cpJson<ListAdversarialCatchesResponse>(
    `/agents/${encodeURIComponent(agentId)}/catches`,
    {
      ...opts,
      allowFallback: opts.allowFixture === true,
      fallback: { items: [] },
    },
  );
}

export async function resolveAdversarialCatch(
  agentId: string,
  catchId: string,
  input: CatchResolutionInput,
  opts: AdversarialCatchesClientOptions = {},
): Promise<AdversarialCatch> {
  return cpJson<AdversarialCatch>(
    `/agents/${encodeURIComponent(agentId)}/catches/${encodeURIComponent(
      catchId,
    )}/resolve`,
    {
      ...opts,
      method: "POST",
      body: input,
      allowFallback: opts.allowFixture === true,
      fallback: localResolvedCatch(agentId, catchId, input),
    },
  );
}
